packages/sku/src/services/vite/plugins/skuViteHMRTelemetry.ts

```diff
@@ -1,45 +1,61 @@
 import type { Plugin } from 'vite';
 import provider from '@/services/telemetry/index.js';
 import debug from 'debug';
+import dedent from 'dedent';
 
 const log = debug('sku:metrics');
 
+// Couldn't quite get the documented method for typing custom payloads working
+// https://vite.dev/guide/api-plugin.html#typescript-for-custom-events
+const customHmrEvent = 'custom:vite-hmr' as const;
+type ViteHmrTimePayload = { durationInMs: number };
+
 export default function skuViteHMRTelemetryPlugin({
   target,
   type,
 }: {
   target: string;
   type: string;
 }): Plugin[] {
-  let startTime = 0;
-
   return [
     {
-      name: 'vite-plugin-sku-hmr-telemetry-pre',
-      enforce: 'pre',
-      handleHotUpdate: {
+      name: 'vite-plugin-sku-hmr-telemetry',
+      transformIndexHtml: {
+        // Vite needs to process the script in order to pick up the HMR context
         order: 'pre',
-        handler() {
-          startTime = performance.now();
-        },
+        handler: () => [
+          {
+            tag: 'script',
+            attrs: { type: 'module' },
+            children: skuHmrTelemetryClient,
+            injectTo: 'head',
+          },
+        ],
       },
-    },
-    {
-      name: 'vite-plugin-sku-hmr-telemetry-post',
-      enforce: 'post',
-      handleHotUpdate: {
-        order: 'post',
-        handler() {
-          const rebuildTime = performance.now() - startTime;
-          log('Rebuild complete: %s', {
-            toString: () => `${Math.round(rebuildTime * 100) / 100}ms`,
-          });
-          provider.timing('start.webpack.rebuild', rebuildTime, {
+      configureServer: (server) => {
+        server.ws.on(customHmrEvent, (data: ViteHmrTimePayload) => {
+          const { durationInMs } = data;
+          log('HMR update completed in %dms', durationInMs);
+          provider.timing('start.webpack.rebuild', durationInMs, {
             target,
             type,
           });
-        },
+        });
       },
     },
   ];
 }
+
+const skuHmrTelemetryClient = dedent/* html */ `
```

Comment:
I think it's possible to point the `src` attribute of the injected `script` to a local file, and vite will inject it into the client. However, I like the idea of having everything related to a plugin in a single file, especially for relatively small plugins like this.